### PR TITLE
feat(vocab): metadata_vocab tag exclude list and count caps (backend)

### PR DESCRIPTION
**Pull request by sydlexius**
## Summary

- Adds a user-configurable tag-filtering layer for genre/style/mood metadata, modeled on MusicBrainz Picard's genre options. A `metadata_vocab` setting holds an `exclude` pattern list (case-insensitive, `*` wildcard) that drops matching tags at tag-merge time, plus per-field count caps (`max_genres`/`max_styles`/`max_moods`).
- Both default to a complete no-op (empty exclude list, zero caps), so existing users see no behavior change until they configure it. Filtering runs in BOTH the orchestrator and the scraper-executor tag-merge paths via the shared pure function `tagdict.ApplyVocabFilter` (a refresh runs through the executor -- see #1589 for the precedent where a one-path wiring silently never ran).
- This is the backend slice (PR 1 of 2); the Settings > Providers > Tag Sources UI follows in a second PR. The design was reworked from an earlier strict-allowlist approach -- see the revised v1 spec in #1130 (comment 4504076245).
- Size override: this PR is 10 hand-written files / ~1.1k LOC, over the 800-LOC guidance, but is already PR 1 of the user-directed 2-PR split and is a cohesive, atomic backend slice (storage + filter logic + dual-path integration + REST API + tests do not partition into independently-working PRs); roughly half the LOC is test code.

## Linked issue

Part of #1130

## Pre-flight checklist

- [x] Pre-push gate green locally (`bash scripts/pre-push-gate.sh`), or run via `/prep-pr` / `/pr-review-toolkit:review-pr` which invokes it.
- [x] Code review pass complete (`/prep-pr` or `/pr-review-toolkit:review-pr`); critical and important findings fixed before pushing.
- [x] Commits squashed into clean, logical commits before the first push (Copilot reviews the diff at PR open).
- [x] At least one label set on `gh pr create --label ...` (the CI label gate fails without one).
- [x] Docs label decision made: `needs-docs-review` for user-visible changes, `docs: not-required` for test, CI, or refactor PRs with no user-visible behavioral change.
- [ ] Screenshot attached for any UI change. (No UI in this slice; the Tag Sources UI is the follow-up PR.)
- [x] UAT performed for user-visible changes (run the binary, not just the test suite).
- [x] OpenAPI spec updated if any request or response shape changed (`make check-openapi`).
- [x] `templ generate` re-run and `*_templ.go` committed if any `.templ` file changed. (No `.templ` files changed.)

## Test plan

- [x] `go test -race ./...` passes locally.
- [x] `bash scripts/pre-push-gate.sh` passes locally (patch coverage 92.6%).
- [x] Manual UAT steps (list the specific flows exercised):
  - [x] `GET /api/v1/settings/vocab` returns the default no-op config (`exclude: []`, zero caps).
  - [x] `PUT` persists an exclude list + caps; round-trips on `GET`; validation rejects unknown fields, blank patterns, and negative caps with 400.
  - [x] With `exclude: ["*grunge"]` set, a refresh of a test artist dropped `post-grunge` (wildcard match) and kept the other four genres -- confirming the filter applies on the executor (refresh) path. Config reset and artist re-refreshed to fully restore.
- [x] Reviewer follow-ups (anything you want a second pair of eyes on):
  - [x] The per-field count cap is applied incrementally (once per provider merge) in the orchestrator; the result matches a single final-pass cap under the existing priority-ordered merge. Reviewed and documented in `vocab.go`.
  - [x] Deferred finding: `getStringSetting` conflates a DB read error with an absent key -- pre-existing shared infra, out of scope here, tracked in #1594.

## Summary by CodeRabbit

* **New Features**
  * Admin API: view and update metadata vocab via GET/PUT /api/v1/settings/vocab (schema validated, strict JSON).
  * Configure exclude patterns (case-insensitive wildcards) and per-field caps for genres/styles/moods.
  * Vocab config is applied during metadata merging to filter and truncate tag lists; unset = no-op.

* **Documentation**
  * OpenAPI schema and operations added for the new settings endpoints.

* **Tests**
  * Comprehensive unit and API tests covering parsing, validation, edge cases, and integration behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/sydlexius/stillwater/pull/1595?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: internal/api/openapi.yaml [hunk 1]
```diff
@@ -30,6 +30,40 @@ components:
...
+        max_styles:
+          type: integer
+          minimum: 0
+          default: 0
+          description: Maximum number of styles to write. 0 means unlimited.
+        max_moods:
+          type: integer
+          minimum: 0
+          default: 0
+          description: Maximum number of moods to write. 0 means unlimited.
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/openapi.yaml:67]: Fixed in e6cbcedb. Added additionalProperties: false to the VocabConfig schema to match the PUT handler DisallowUnknownFields contract.

File: internal/api/handlers_vocab.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	dec := json.NewDecoder(req.Body)
+	dec.DisallowUnknownFields()
+
+	var cfg tagdict.VocabConfig
+	if err := dec.Decode(&cfg); err != nil {
+		// Keep the client-facing message generic (consistent with the other
+		// settings handlers); the decoder error detail is not surfaced.
+		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid request body"})
+		return
+	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/handlers_vocab.go:83]: Fixed in e6cbcedb. The PUT handler now does a second decode and rejects anything other than io.EOF, so trailing JSON after the first object returns 400.

File: internal/provider/tagdict/vocab.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+func WithMetadataVocab(ctx context.Context, cfg *VocabConfig) context.Context {
+	return context.WithValue(ctx, ctxKeyMetadataVocab{}, cfg)
+}
+
+// MetadataVocab retrieves the VocabConfig from the context. Returns nil when
+// no config has been injected; ApplyVocabFilter treats nil as a no-op config.
+func MetadataVocab(ctx context.Context) *VocabConfig {
+	cfg, _ := ctx.Value(ctxKeyMetadataVocab{}).(*VocabConfig)
+	return cfg
+}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/provider/tagdict/vocab.go:90]: Fixed in e6cbcedb. Both helpers now guard a nil context: WithMetadataVocab falls back to context.Background, MetadataVocab returns nil.

File: internal/api/handlers_vocab.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+		r.logger.Error("marshaling metadata_vocab config", "error", err)
+		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "internal error"})
+		return
+	}
+
+	now := time.Now().UTC().Format(time.RFC3339)
+	_, err = r.db.ExecContext(req.Context(),
+		`INSERT INTO settings (key, value, updated_at) VALUES (?, ?, ?)
+		ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at`,
+		SettingMetadataVocab, string(blob), now)
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/handlers_vocab.go:135]: Fixed in e6cbcedb. handlePutVocab normalizes a nil Exclude to an empty slice before marshaling, so the persisted blob is never exclude:null.

File: internal/api/handlers_vocab_test.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+	body := `{"max_genres":-1}`
+	req := httptest.NewRequest(http.MethodPut, "/api/v1/settings/vocab", strings.NewReader(body))
+	w := httptest.NewRecorder()
+	r.handlePutVocab(w, req)
+
+	if w.Code != http.StatusBadRequest {
+		t.Fatalf("expected 400 for negative cap, got %d: %s", w.Code, w.Body.String())
+	}
+}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/handlers_vocab_test.go:168]: Fixed in e6cbcedb. Added TestHandlePutVocab_TrailingJSON asserting a 400 for a valid object followed by trailing JSON.

File: internal/scraper/executor_test.go [hunk 6]
```diff
@@ -1364,3 +1365,43 @@ func TestMembersFieldQueried(t *testing.T) {
...
+			MetadataVocabCfg: &tagdict.VocabConfig{MaxGenres: 2},
+		}
+		if !fieldAppliers[FieldGenres](m, r) {
+			t.Fatal("FieldGenres applier should report the field populated")
+		}
+		if len(r.Metadata.Genres) != 2 {
+			t.Fatalf("executor path did not apply the count cap: %v", r.Metadata.Genres)
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/scraper/executor_test.go:1439]: Fixed in e6cbcedb. TestFieldAppliers_VocabFilter is now table-driven over genres, styles, and moods for both the exclude and cap cases.

File: internal/provider/orchestrator_test.go [hunk 7]
```diff
@@ -3254,3 +3256,42 @@ func TestFetchMetadata_MembersAuthoritative(t *testing.T) {
...
+			Metadata:         &ArtistMetadata{},
+			MetadataVocabCfg: &tagdict.VocabConfig{MaxGenres: 2},
+		}
+		pr := &providerResult{meta: &ArtistMetadata{Genres: []string{"Rock", "Pop", "Jazz", "Blues"}}}
+
+		applyTagSliceField(result, "genres", pr, NameMusicBrainz)
+
+		if len(result.Metadata.Genres) != 2 {
+			t.Fatalf("orchestrator path did not apply the count cap: %v", result.Metadata.Genres)
+		}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/provider/orchestrator_test.go:3296]: Fixed in e6cbcedb. The cap subtest now asserts the exact retained genres [Rock, Pop], not just the length.

File: internal/api/openapi.yaml [hunk 8]
```diff
@@ -5739,6 +5773,78 @@ paths:
...
+    get:
+      tags: [Settings]
+      summary: Get metadata vocabulary configuration
+      operationId: getVocabConfig
+      description: >
+        Returns the current metadata_vocab configuration, which controls
+        per-field allowlists, mapping rules, and optional strict-filter mode
+        for genres, styles, and moods. When the setting has never been saved,
+        the response is the default seed config (strict=false for all fields,
+        which is a no-op).
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/openapi.yaml]: Fixed in e6cbcedb. Rewrote the GET and PUT /settings/vocab endpoint descriptions to the exclude-list + count-cap shape; the strict/allowlist/mapping wording was stale.

File: internal/scraper/executor_test.go [hunk 9]
```diff
@@ -1364,3 +1365,74 @@ func TestMembersFieldQueried(t *testing.T) {
...
+			f.set(m, []string{"Rock", "Pop", "Jazz", "Blues"})
+			cfg := &tagdict.VocabConfig{}
+			f.setCap(cfg, 2)
+			r := &provider.FetchResult{Metadata: &provider.ArtistMetadata{}, MetadataVocabCfg: cfg}
+			if !fieldAppliers[f.name](m, r) {
+				t.Fatalf("%s applier should report the field populated", f.name)
+			}
+			if got := f.get(r.Metadata); len(got) != 2 {
+				t.Fatalf("%s: executor path did not apply the count cap: %v", f.name, got)
+			}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/scraper/executor_test.go:1436]: Fixed in 3f6e55c1. The executor cap subtest now asserts the exact retained tags [Rock, Pop], not just the length.
